bundle/edu.gemini.ictd/src/main/scala/edu/gemini/ictd/FeatureAvailability.scala

```diff
@@ -0,0 +1,155 @@
+
+  type AvailabilityMap = Map[java.lang.Enum[_], Availability]
+
+  val EmptyAvailabilityMap: AvailabilityMap =
+    Map.empty
+
+  sealed trait InstrumentAvailability {
+    def instrument: Instrument
+
+    def availabilityMap: AvailabilityMap
+  }
+
+  /** Creates a FeatureAvailability instance given the collection of every
+    * instrument's FeatureTables from the ICTD database.
+    */
+  def fromTables(tabs: Map[Instrument, FeatureTables]): FeatureAvailability = {
+    val forInst = tabs.withDefaultValue(FeatureTables.empty)
+
+    FeatureAvailability(
+      Flamingos2Config.fromTables(forInst(Flamingos2)),
+      GmosNorthConfig.fromTables(forInst(GmosNorth)),
+      GmosSouthConfig.fromTables(forInst(GmosSouth))
+    )
+  }
+
+  final case class Flamingos2Config(
+    filters: Map[F2Filter, Availability],
+    fpus:    Map[F2FPUnit, Availability]
+  ) extends InstrumentAvailability {
+
+    def instrument: Instrument =
+      Instrument.Flamingos2
+
+    def availabilityMap: AvailabilityMap =
+      filters ++ fpus
+
+  }
+
+  object Flamingos2Config {
+    def fromTables(tabs: FeatureTables): Flamingos2Config =
+      Flamingos2Config(
+        resolve[F2Filter](tabs.filter),
+        resolve[F2FPUnit](tabs.longslit ++ tabs.circular)
+      )
+  }
+
+
+  final case class GmosNorthConfig(
+    dispersers: Map[DisperserNorth, Availability],
+    filters:    Map[FilterNorth,    Availability],
+    fpus:       Map[FPUnitNorth,    Availability]
+  ) extends InstrumentAvailability {
+
+    def instrument: Instrument =
+      Instrument.GmosNorth
+
+    def availabilityMap: AvailabilityMap =
+      dispersers ++ filters ++ fpus
+
+  }
+
+  object GmosNorthConfig {
+    def fromTables(tabs: FeatureTables): GmosNorthConfig =
+      GmosNorthConfig(
+        resolve[DisperserNorth](tabs.grating),
+        resolve[FilterNorth   ](tabs.filter),
+        resolve[FPUnitNorth   ](tabs.ifu ++ tabs.longslit ++ tabs.nsLongslit)
+      )
+  }
+
+  final case class GmosSouthConfig(
+    dispersers: Map[DisperserSouth, Availability],
+    filters:    Map[FilterSouth,    Availability],
+    fpus:       Map[FPUnitSouth,    Availability]
+  ) extends InstrumentAvailability {
+
+    def instrument: Instrument =
+      Instrument.GmosSouth
+
+    def availabilityMap: AvailabilityMap =
+      dispersers ++ filters ++ fpus
+
+  }
+
+  object GmosSouthConfig {
+    def fromTables(tabs: FeatureTables): GmosSouthConfig =
+      GmosSouthConfig(
+        resolve[DisperserSouth](tabs.grating),
+        resolve[FilterSouth   ](tabs.filter),
+        resolve[FPUnitSouth   ](tabs.ifu ++ tabs.longslit ++ tabs.nsLongslit)
+      )
+  }
+
+  // Given a map of String key to Availability drawn from the ICTD, and an Enum
+  // type A, find the availability of each instance of A.
+  private def resolve[A <: java.lang.Enum[A] with IctdType : ClassTag](
+    tab: AvailabilityTable
+  ): Map[A, Availability] = {
+
+    val c  = classTag[A].runtimeClass.asInstanceOf[Class[A]]
```

Comment:
Reflection is often problematic. Do you see any risks here?

Reply:
If you had a typeclass instance for each `A` that provided the values you would't need to use a tag, but this is probably good enough. We're not looking things up by name, which is where you can get into trouble.